src/coreComponents/physicsSolvers/multiphysics/ThermalSinglePhasePoromechanicsKernel.hpp

```diff
@@ -0,0 +1,606 @@
+      1.0,
+      detJxW );
+
+    BilinearFormUtilities::compute< pressureTestSpace,
+                                    pressureTrialSpace,
+                                    DifferentialOperator::Identity,
+                                    DifferentialOperator::Identity >
+    (
+      stack.dLocalResidualEnergy_dTemperature,
+      1.0,
+      stack.dFluidEnergyIncrement_dTemperature,
+      1.0,
+      detJxW );
+  }
+
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void quadraturePointKernel( localIndex const k,
+                              localIndex const q,
+                              StackVariables & stack ) const
+  {
+    // Step 1: compute displacement finite element basis functions (N), basis function derivatives (dNdX), and
+    // determinant of the Jacobian transformation matrix times the quadrature weight (detJxW)
+    real64 N[numNodesPerElem]{};
+    real64 dNdX[numNodesPerElem][3]{};
+    FE_TYPE::calcN( q, N );
+    real64 const detJxW = m_finiteElementSpace.template getGradN< FE_TYPE >( k, q, stack.xLocal, dNdX );
+
+    // Step 2: compute strain increment
+    real64 strainIncrement[6]{};
+    FE_TYPE::symmetricGradient( dNdX, stack.uhat_local, strainIncrement );
+
+    // Step 3: compute 1) the total stress, 2) the body force terms, and 3) the fluidMassIncrement
+    // using quantities returned by the PorousSolid constitutive model.
+    // This function also computes the derivatives of these three quantities wrt primary variables
+    smallStrainUpdate( k, q, strainIncrement, stack );
+
+    // Step 4: use the total stress and the body force to increment the local momentum balance residual
+    // This function also fills the local Jacobian rows corresponding to the momentum balance.
+    assembleMomentumBalanceTerms( N, dNdX, detJxW, stack );
+
+    // Step 5: use the fluid mass increment to increment the local mass balance residual
+    // This function also fills the local Jacobian rows corresponding to the mass balance.
+    assembleElementBasedFlowTerms( dNdX, detJxW, stack );
+  }
+
+  /**
+   * @copydoc geosx::finiteElement::ImplicitKernelBase::complete
+   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  real64 complete( localIndex const k,
+                   StackVariables & stack ) const
+  {
+    real64 const maxForce = Base::complete( k, stack );
+
+    constexpr integer nUDof = numNodesPerElem * numDofPerTestSupportPoint;
+
+    // Step 1: assemble the derivatives of linear momentum balance wrt temperature into the global matrix
+
+    for( integer localNode = 0; localNode < numNodesPerElem; ++localNode )
+    {
+      for( integer dim = 0; dim < numDofPerTestSupportPoint; ++dim )
+      {
+        localIndex const dof = LvArray::integerConversion< localIndex >( stack.localRowDofIndex[numDofPerTestSupportPoint*localNode + dim] - m_dofRankOffset );
+        if( dof < 0 || dof >= m_matrix.numRows() ) continue;
+
+        m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
```

Comment:
is it possible to use addToRow here instead of addToRowBinarySearchUnsorted as we are adding a single value?